### ISSUE TITLE
bundle install is not installing project vendor cached git gems, but using the cache as the install location

### ISSUE THREAD

**Issue by bughit** [rubygems/rubygems#3204]
My understanding is that the cache is only supposed to be used as a source for installation of gems, and not function as the install location, from which the code is ultimately loaded.

The normal git gem system install location is `~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/bundler/`, so when I bundle install (normally, without --path or --deployment) a gem that is cached in the project's vendor/cache I expect it to end up in `~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/bundler/`

It doesn't, and `bundle show` shows its location in `$PROJECT/vendor/cache/`.

If it's removed from  `$PROJECT/vendor/cache/`  and bundle reinstalled, it does get installed to `~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/bundler/` but `bundle show` still shows `$PROJECT/vendor/cache/`

Is this expected behavior?  Does not seem like a good idea.

**Comment by dentarg** [comment]
I just ran into this, so I checked if it was still a problem with the latest released Bundler, 2.1.4 (and RubyGems 3.1.2), and it is. [Recorded the steps](https://github.com/dentarg/puma-vagrant/commits/master/vendoring_master_app)  to reproduce [in git](https://github.com/dentarg/puma-vagrant/tree/a83aa96eabb6a316636b412ef0cc6f84f06e0c58/vendoring_master_app).

I understand that this problem is somewhat of a niche use-case, but just wanted to post some up-to-date info on this, seeing the issue hasn't received any comments in three years.

**Comment by thedanbob** [comment]
I've been banging my head against this bug for two days. Running a second `bundle install` doesn't fix the problem for me. I've only found two ways to properly install a git gem in deployment mode:

1. Manually create a .gem file and copy it into vendor/cache, then run `bundle install`
2. Manually copy the gem directory from vendor/cache to vendor/bundle/ruby/2.x.0/bundler/gems/

Oddly enough I don't have a problem with bundler 1.x, `bundle install --deployment --local` installs git gems correctly for me. Only 2.1.4 is broken.

**Comment by deivid-rodriguez** [comment]
I had a quick look. Could you try [this patch](https://github.com/rubygems/rubygems/compare/try_fix_git_cache_issue) and see if it fixes the issue?

**Comment by thedanbob** [comment]
You're my hero! `bundle install` and `bundle package` work as expected with that patch.

**Comment by deivid-rodriguez** [comment]
Awesome! Checking CI now, I will try to write a spec tomorrow if everything is good.

**Comment by deivid-rodriguez** [comment]
Actually, even if my patch seems to fix things, I think it's doing it the wrong way since the way it "fixes this" is to make sure git gems are actually fully installed into the cache location, instead of making sure git gems are not looked up in the cache location but in the real install location. I'll try to fix it the right way.

**Comment by bughit** [comment]
@deivid-rodriguez 

> instead of making sure git gems are not looked up in the cache location but in the real install location

That's the second part, the first is for install them to the appropriate location. That was probably implied by you, but can't hurt to make it explicit.

**Comment by deivid-rodriguez** [comment]
I think the gem is actually installed to the proper location, and then copied to the cache. The problem is that subsequent `bundle exec` invocations will use the cache as the install location.

But to be completely honest I only went back to https://github.com/rubygems/rubygems/issues/3204#issuecomment-313294265 where you provided a zip with a repro, and didn't look at older messages, so maybe there's more bugs here.

**Comment by thedanbob** [comment]
In my case the gems were not getting installed where they should have been (vendor/bundle/ruby/2.x.0/bundler/gems) until your patch. Maybe they were getting "installed" to the cache?

**Comment by bughit** [comment]
> I think the gem is actually installed to the proper location

When I reported this they definitely weren't. I'll try the current version at some point.

**Comment by deivid-rodriguez** [comment]
> In my case the gems were not getting installed where they should have been (vendor/bundle/ruby/2.x.0/bundler/gems) until your patch. Maybe they were getting "installed" to the cache?

Interesting, maybe. For now I'm focusing on @bughit's repro case, but I'll try to look at yours too.

In the case of git gems, "getting installed" vs "getting cached" is a quite subtle distinction. One difference is "installing" needs to compile extensions while caching doesn't. With my patch, the extension started being compiled inside the cache folder, which seems unexpected, that's how I noticed the problem.

**Comment by bughit** [comment]
You should check both scenarios, git gem with/without native extensions.

When reported:

- without extension
  - install was a noop
  - gem was loaded from cache
- with extension
 - just the extension was installed outside of the cache (don't remember if it was always the right place)
 - gem loaded from cache

expected:
 
- git gem install should always go the specified bundle location 
- git gems should never load from the vendor cache, always from the bundle

**Comment by thedanbob** [comment]
I was wrong, even with the patch `bundle install` doesn't work as expected if the git gem is already cached.

> I think the gem is actually installed to the proper location, and then copied to the cache. The problem is that subsequent bundle exec invocations will use the cache as the install location.

This appears to be correct, if vendor/cache doesn't exist the git gem is installed in the correct location and also appears in the cache. But if it's already present in vendor/cache it doesn't get installed in vendor/bundle.

**Comment by deivid-rodriguez** [comment]
Thanks for reporting your testing about current behaviour @thedanbob, hopefully I'll be able to return to this ticket soon!

**Comment by rocketraman** [comment]
Just ran into this with Ruby 2.7.0 and Puma 5.0.0. I've read through the above but not being a Ruby guy, I don't really understand what the correct workaround is. Can anyone ELI5?

**Comment by rocketraman** [comment]
Solved it by updating my Gemfile to Puma 5.1.1.

**Comment by jeremy** [comment]
Discovered this while trying to use a Docker volume mount as a persistent Bundler cache with Bundler 2.2.9.

Dockerfile excerpt:
```
RUN --mount=id=bundler-cache,type=cache,target=/bundler-cache \
    echo -n "Cache size pre-install: " && du -sh /bundler-cache && \
    echo "Configuring Bundler to use persistent gems/git cache for installs:" && \
      bundle config set cache_path /bundler-cache && \
      bundle config set cache_all true && \
      bundle config set no_prune true && \
    echo "Bundling from cache: bundle install --local" && \
      bundle install --local && \
    echo -n "Cache size post-install: " && du -sh /bundler-cache && \
    echo -n "Bundle size: " && du -sh "$BUNDLE_PATH" && \
    echo -n "Checking bundle with cache volume mounted: " && bundle check && \
    bundle config unset cache_path && \
    echo -n "Checking bundle after changing cache_path: " && bundle check

RUN echo -n "Checking bundle with cache volume unmounted: " && bundle check
```

The `bundle check` fails after changing the cache path or unmounting the Docker volume. Git gems are installed using the cache dir, so removing the cache dir breaks the bundle.

Good luck y'all!

**Comment by cjlarose** [comment]
FWIW I ran into this issue when deploying a lambda function to AWS. I was packaging the `vendor/bundle` directory for distribution to the lambda function, but not the `vendor/cache` directory. The fix isn't as easy as just adding `vendor/cache` to the lambda function, though.

At first I thought the way this worked at all was that normally Bundler would add the relevant subdirectories of `vendor/cache` to `GEM_PATH` so that when I `require` that gem, Rubygems can find it. That's not true. AFAICT, the only way you can actually use gems (and observe the behavior where Bundler loads gems from the `vendor/cache` directory) is if you're using `Bundler.require` to require the gems. If you're _not_ using `Bundler.require` and just using good ol' Rubygems `require`, the gem will not be loaded at all (`require` fails).

**Comment by thedanbob** [comment]
This works perfectly for me. And I agree with your reasoning, the workflow my org uses for deployment is

1. Cache gems locally
2. Push `vendor/cache` and `Gemfile*` to a "build server" with tools for compiling extensions, git, etc.
3. Copy the resulting `vendor/bundle` (and `vendor/cache`, because of this issue) to production

The new behavior makes this work without copying `vendor/cache` to production.

**Comment by deivid-rodriguez** [comment]
Thanks for trying it @thedanbob! The attached PR should be ready now :+1:.

**Comment by bughit** [comment]
@deivid-rodriguez 

What is git going to be used for during bundle install?

- The vendor cache does not include the git repo.
- `--local` is supposed to install from cache and not hit the network, otherwise the cache is pointless

So if you actually install from cache, which it's not clear bundler can do at this point, since by default it does git operations, then you don't need git. And if you do need git, it sounds like you're not using the cache and violating `--local`

**Comment by bughit** [comment]
> The vendor cache does not include the git repo.

As well it shouldn't.

**Comment by deivid-rodriguez** [comment]
You seem to be making the assumption that the only purpose bundler uses git for is to do remote operations, which is not the case. If you can spot any actual problems with the patch, please point to them including reproducible examples. Thanks.

**Comment by bughit** [comment]
I am making no assumptions, but asking you a question, did you miss the question mark? Which you are not answering for some reason, why?

If you are installing from the vendor cache which does not have the git repo, what is git needed for?

**Comment by deivid-rodriguez** [comment]
You're not being helpful, and not willing to try anything, even if this issue was reported by you originally. If you're still interested in this issue, please temporary remove git from your PATH and try for yourself. If you're no longer interested in this issue, you don't need any answers.

**Comment by thedanbob** [comment]
For what it's worth he's right, git gems aren't installed from the cache but are freshly cloned even with `--local`. These are the errors I got when I moved git out of my PATH:

```
Fetching git@<redacted>.git

Retrying `git clone git@<redacted>.git /home/dan/projects/current/<redacted>/vendor/bundle/ruby/2.5.0/cache/bundler/git/<redacted> --bare --no-hardlinks --quiet` at /home/dan/projects/current/<redacted> due to error (2/4): Errno::ENOENT No such file or directory - git                                                                                                              
```

**Comment by bughit** [comment]
I am trying to understand how things are going to work with your fix, which you are refusing to explain, and with completely unreasonable hostility.

You said git is now a dependency, so it's perfectly reasonable to understand why. Obviously it doesn't make sense to force such a dependency unless it's actually needed.

If you are installing from the vendor cache which does not have a git repo, what is git for?

**Comment by deivid-rodriguez** [comment]
Yeah, @thedanbob, I was hoping my patch might fix [the issue](https://github.com/rubygems/rubygems/issues/4226), but it doesn't :(. In any case, that issue already exists, so it seems independent from the patch.

**Comment by deivid-rodriguez** [comment]
Since I picked up this issue I felt like you've been making a lot of questions, without actually willing to try anything for yourself, giving me the impression that you were demanding things from me without even being interested anymore in the original issue. Anyways, I admit my previous message was too hostile, I apologize.

**Comment by bughit** [comment]
> I felt like you've been making a lot of questions, without actually willing to try anything for yourself

I was making the necessary clarifying comments to make sure the problem was properly understood and so properly addressed. How you can find fault with that is beyond comprehension.

> Anyways, I admit my previous message was too hostile, I apologize.

Ok. Let's move beyond that and stick to technical considerations.

I would argue that the tradeoff this fix introduces for `--local`, which is the main use-case for the vendor cache:

- requiring git
- requiring the network
- bypassing the cache

makes things worse. It introduces performance and security issues that `--local` was intended to avoid, and which weren't present before the fix. So the fix should be backed out until install from cache works.

**Comment by deivid-rodriguez** [comment]
@thedanbob Can you confirm whether this need for `git` to `bundle install --local` is not something caused by my patch? I'd say it is not, specially since we already had that issue I linked to reported not long ago, but I'd like to confirm it.

If my PR is introducing a bug with `bundle install --local`, I'm happy to revert it of course

**Comment by thedanbob** [comment]
Actually I'm not sure now how the other guy ran into this. Perhaps in the `requires_checkout?` call, which checks the local revision against the remote. Technically this patch does also result in the same issue, since before bundler was treating git gems as already installed once they're in the cache, but now they aren't and thus must be installed with a fresh clone.

**Comment by deivid-rodriguez** [comment]
Aaaaaah, I see, you too are indeed right :facepalm:. I'll have a closer look at this tomorrow, tricky stuff this issue :sweat_smile:.

**Comment by bughit** [comment]
> before bundler was treating git gems as already installed once they're in the cache, but now they aren't and thus must be installed with a fresh clone.

The pre-patch behavior, although buggy, is at least consistent with the intent of  `--local`. And having `--local` actually use the cache and be local is more important.

**Comment by deivid-rodriguez** [comment]
Agreed, breaking `--local` when you have a cache would make things worse since it would completely defeat the purpose of having a cache. If I find a solution quickly I'll apply it, otherwise I'll revert the patch.

**Comment by deivid-rodriguez** [comment]
Something I don't quite understand is that I explicitly changed a spec to test exactly the behaviour we expect. Run `bundle cache`, then run `bundle install --local` and make sure no "Fetching" happens. Anyways, it's bed time for me, hopefully I'll get answers soon.

**Comment by thedanbob** [comment]
I figured out why the test didn't fail: `bundle cache --no-install` actually *does* install git gems (and only git gems), and to the correct location of vendor/bundle/ruby/2.5.0/bundler/gems/. So there's no "Fetching' in the later `bundle install` because it's already installed!

I guess I should open another issue for `bundle cache` not respecting the --no-install flag for git gems 😂

**Comment by bughit** [comment]
I've been focusing on the --local  case as it's the most important, however it's worth noting that even without --local, if the required git gem version is in the vendor cache it has to be installed from there. The only exception to that (which proves the rule) is:

> [By default, if you run bundle install after running bundle cache(1), bundler will still connect to rubygems.org to check whether a platform-specific gem exists for any of the gems in vendor/cache](https://bundler.io/man/bundle-cache.1.html)

But this doesn't AFAICT apply to git gems.

Now to the git gem install from vendor cache process.

The default git gem install process, as I understand it, is as follows (the paths are from memory so could be off):

- the git repo is bare cloned into `gems/${VER}/cache/bundler/git/${GEM_NAME}`
  - future fetches happen here
- the bare clone is again cloned (non bare) to the ultimate install location: `gems/${VER}/bundler/${GEM_NAME}_${COMMIT_HASH}/`
  - and the required commit is checked out
  - from a prior conversation with bundler people, the git repo in the install location is not needed and not used and I suggested should be deleted after checkout

This obviously does not apply when installing from the vendor cache, which would be:

- avoiding the git dependency
- copy the gem folder from the vendor cache to `gems/${VER}/bundler/`
- it will not contain the git repo, which is fine because as I mentioned it's not actually used

**Comment by deivid-rodriguez** [comment]
> I figured out why the test didn't fail: bundle cache --no-install actually does install git gems (and only git gems), and to the correct location of vendor/bundle/ruby/2.5.0/bundler/gems/. So there's no "Fetching' in the later bundle install because it's already installed!

That makes sense, I'll have another look at this now!

**Comment by deivid-rodriguez** [comment]
@thedanbob Can you try #4469 now?

I think it should fix the reported issues with `bundle install --local`. The problem was that `bundle cache` was removing git information from the cache location, which made it impossible to figure out whether the cache had the proper information for the install to succeed, or needed to fetch from the remote repository. I now changed things so that the "per-app" cache works just like the "per-ruby" cache. I would expect that to work just fine.

**Comment by thedanbob** [comment]
I think it's working! Here's what I tried:
1. `bundle cache --no-install`
2. To simulate the remote repo being unavailable, changed the git gem's remote URL to something invalid in Gemfile, Gemfile.lock, and cached clone
3. `bundle install --local --deployment` (succeeded)
4. Deleted bundle/cache directory
5. `bundle exec rails console` (succeeded)
6. `bundle info <git gem>` shows the correct install path

**Comment by deivid-rodriguez** [comment]
That's great to hear. Hopefully this patch is finally ok :pray:.

**Comment by thedanbob** [comment]
It should be noted somewhere that this is a backwards-incompatible change. A `bundle cache` with this patch will break a `bundle install --local` or `bundle update` with an earlier version of bundler, since the cached gem directory is now a bare git repo rather than a clone.

**Comment by deivid-rodriguez** [comment]
And I guess without https://github.com/rubygems/rubygems/pull/4469/commits/9e8071b5881f463018a95ad1361d0de3a85be04d, those commands on a cache generated without older versions also fail?

I think I'd be fine with forcing people using caches for git repositories to regenerate their caches for the sake of fixing this bug. But I guess this could be made backwards compatible with a bit more work, I'll think of how hard that'd be.

**Comment by bughit** [comment]
> since the cached gem directory is now a bare git repo rather than a clone.

@deivid-rodriguez Shouldn't this get some discussion and feedback?

In my view that's highly undesirable. I don't want to be needlessly caching and thus storing in the project repo multiple whole gem repos and packaging/transmitting them, when all you need is a single commit. The install from cache can be: "copy the git gem folder from the vendor cache to install location". Is that not true?

**Comment by indirect** [comment]
That is not true. Git gems include a git repo because many gems contain code that relies on the presence of a git repo to build and run from a checkout (eg gemspecs that shell out to git). Simply copying the gem's files out of the git repo often produces broken gems.

**Comment by thedanbob** [comment]
It does seem unnecessary to clone the entire repo. If the cache needs to be a bare repo to make things work properly, could it not be a shallow clone? e.g. `git clone --bare --depth 1 <repo>`

**Comment by dentarg** [comment]
>A gemspec that shells out to git and does what?

List files is common, example: https://github.com/socketry/nio4r/blob/05532a5a35758c812498a1767c25118a7f64e101/nio4r.gemspec#L16

**Comment by bughit** [comment]
> List files is common

Well, if install needs to do this, the shallow clone idea could possibly cover this, although the mechanism may need to be different (via fetch)

its possible to fetch a specific commit:
https://stackoverflow.com/a/43136160

I tried it with this repo

```
git init
git remote add origin https://github.com/rubygems/rubygems.git
git fetch --depth 1 origin 7b9bcf8027ee20c8e6ae2389988e9305b0adbe2f
git checkout FETCH_HEAD
```

**Comment by pbhogan** [comment]
First, many thanks to @deivid-rodriguez for all his hard work on solving this issue. It's been a real pain requiring ugly hacks for clean docker deployments.

Second, I think the issue of cloning the whole repo is not a huge deal. Gems repos are not typically large, and it's fairly rare to have many git gems in a project. Even some of our largest projects which have a ton of gems typically have only 4 or 5 git gems. I looked up one of our projects with the most git gems and it came to a grand total of... 700 KB.

Of course, if you had Rails edge cloned, that would get big. But that's an outlier. If a shallow clone could do the trick, fantastic. That'd be a nice savings for what is really an edge case, and be slightly faster in general for initial bundles. Homebrew recently switched to doing shallow clones as a time/space saving measure too.

But if not, just having the actual issue solved is a vast improvement over what we had before.

**Comment by bughit** [comment]
> If a shallow clone could do the trick, fantastic.

The actual shallow clone operation does not appear to be up to the task. It clones the latest commit(s) up to the `--depth` but you need an arbitrary commit, specified in the lock file. The shallow fetch on the other hand should do the trick: `git fetch --depth 1 origin 7b9bcf8027ee20c8e6ae2389988e9305b0adbe2f`

**Comment by deivid-rodriguez** [comment]
I agree with @pbhogan, I think the current solution is already an improvement, specially since the standard "per ruby installation" cache already clones whole bare repositories, and that current workarounds for git gem caching involve also copying over the install folder, which also currently includes the whole repo history.

That said, I'll spend some time too to see if I can make git repo cloning lighter in general.

**Comment by searls** [comment]
It seems likely that with the [ongoing mimemagic issue](https://github.com/rails/rails/issues/41750), a lot of folks may be looking to bundler to vendor _only_ mimemagic as opposed to all of their gems and may run into this issue as a result.

**Comment by deivid-rodriguez** [comment]
I worked on improving repo cloning with the suggested `git fetch` option at https://github.com/rubygems/rubygems/pull/4475. I does show significant improvements, I'm just worried about compatibility with old gits and newer gits without the necessary feature enabled.

**Comment by cornfeedhobo** [comment]
This is sadly still an issue. Outsider's question; why is this not as easy as dropping the cloned repo into `gems/`? Why are these dependencies getting stuck in `cache/`?

**Comment by stevenharman** [comment]
I just hit the same thing - I was trying to `rm -rf ./vendor/cache` to save some space in a Docker image after installing via:

```console
bundle config set --local deployment "true"
bundle config set --local path "vendor/bundle"
bundle check || bundle install --local --jobs=$CPU_COUNT --retry=3
```

But that makes it look like git Gems aren't installed.

@deivid-rodriguez It looks like the fix you posted (https://github.com/rubygems/rubygems/pull/4469) was never actually merged? So is this still an open and active issue?

**Comment by deivid-rodriguez** [comment]
@deivid-rodriguez Sorry for not answering earlier. I'm about to merge the fix, can you try it?

**Comment by miry** [comment]
I came here for same reason as @stevenharman.
The hint about location of the git gem came from `bundle install --prefer-local --verbose`.

```
#24 424.2 Using somegem 2.9.8 from https://github.com/user/somegem.git (at /app/vendor/cache/somegem-0ee1e7b22250@0ee1e7b)
#24 424.2 0:  somegem (2.9.8) from /app/vendor/cache/somegem-0ee1e7b22250/somegem.gemspec
```

Version:

```
$ bundle version
Bundler version 2.5.14 (2024-06-21 commit 72920bc230)
```
